source/extensions/filters/http/ratelimit/ratelimit.cc

```diff
@@ -55,52 +55,67 @@ void Filter::initiateCall(const Http::RequestHeaderMap& headers) {
     return;
   }
 
-  Router::RouteConstSharedPtr route = callbacks_->route();
-  if (!route || !route->routeEntry()) {
-    return;
+  std::vector<Envoy::RateLimit::Descriptor> descriptors;
+  populateRateLimitDescriptors(descriptors, headers, false);
+  if (!descriptors.empty()) {
+    state_ = State::Calling;
+    initiating_call_ = true;
+    client_->limit(*this, getDomain(), descriptors, callbacks_->activeSpan(),
+                   callbacks_->streamInfo(), getHitAddend());
+    initiating_call_ = false;
   }
+}
 
-  cluster_ = callbacks_->clusterInfo();
-  if (!cluster_) {
-    return;
+void Filter::populateRateLimitDescriptors(std::vector<Envoy::RateLimit::Descriptor>& descriptors,
+                                          const Http::RequestHeaderMap& headers,
+                                          bool on_stream_done) {
+  // To use the exact same context for both request and on_stream_done rate limiting descriptors,
+  // we save the route and per-route configuration here and use it later.
+  if (!on_stream_done) {
+    route_ = callbacks_->route();
+    if (!route_) {
+      return;
+    }
```

Comment:
move this check out of if to ensure the null route_ will not crash envoy when on_stream_done is true.